### PR TITLE
fix: add connectionInfo to telemetry calls, and connectionId to events COMPASS-7968

### PR DESCRIPTION
**Pull request by paula-stacho**
### strategy

- we only send connectionId, but pass the whole connectionInfo to telemetry to be more future-ready
- if available, ConnectionInfoAccess is used
- otherwise, ConnectionRepository is used - this would be mainly for these plugins https://github.com/mongodb-js/compass/blob/cc3acaf760854e7e8d3dc3dc56ea11aeaff2c345/packages/compass/src/app/components/workspace.tsx#L108-L113

### event plan

These are the events that I added connectionInfo to

  - 'Aggregation Canceled'
  - 'Aggregation Edited'
  - 'Aggregation Executed'
  - 'Aggregation Explained'
  - 'Aggregation Export Opened'
  - 'Aggregation Exported'
  - 'Aggregation Opened'
  - 'Aggregation Saved'
  - 'Aggregation Saved As View'
  - 'Aggregation Side Panel Opened'
  - 'Aggregation Timed Out'
  - 'Aggregation Use Case Added'
  - 'Aggregation Use Case Saved'
  - 'AI Prompt Submitted'
  - 'AI Query Feedback'
  - 'AI Response Failed'
  - 'AI Response Generated'
  - 'Bulk Delete Executed'
  - 'Bulk Delete Opened'
  - 'Bulk Update Executed'
  - 'Bulk Update Favorited'
  - 'Bulk Update Opened'
  - 'Collection Created'
  - 'Connection Attempt'
  - 'Connection Failed'
  - 'CurrentOp showOperationDetails'
  - 'Database Created'
  - 'Delete Export Opened'
  - 'Delete Exported'
  - 'DetailView hideOperationDetails'
  - 'DetailView killOp'
  - 'Document Cloned'
  - 'Document Copied'
  - 'Document Deleted'
  - 'Document Inserted'
  - 'Document Updated'
  - 'Editor Type Changed'
  - 'Explain Plan Executed'
  - 'Export Completed'
  - 'Export Opened'
  - 'Query Edited'
  - 'Query Executed'
  - 'Query Export Opened'
  - 'Query Exported'
  - 'Query History Closed'
  - 'Query History Favorite Added'
  - 'Query History Favorites'
  - 'Query History Opened'
  - 'Query History Recent'
  - 'Query History Recent Used'
  - 'Query Results Refreshed'
  - 'Import Completed'
  - 'Import Error Log Opened'
  - 'Import Opened'
  - 'Index Create Opened'
  - 'Index Created'
  - 'Index Dropped'
  - 'Index Edited'
  - 'New Connection'
  - 'Performance Advisor Clicked'
  - 'Performance Paused'
  - 'Performance Resumed'
  - 'PipelineAI Feedback'
  - 'Schema Analyzed'
  - 'Schema Validation Added'
  - 'Schema Validation Edited'
  - 'Schema Validation Updated'
  - `Shell ${string}`
  - 'Switch View Type'
  - 'Update Export Opened'
  - 'Update Exported'
  - 'View Updated'
  - 'Focus Mode Closed'
  - 'Focus Mode Opened'

**These have the connectionInfo sometimes**

  - 'Query History Favorite Removed' - depends! this one can be from 'my_queries' or from 'documents' screen
  - 'Query History Favorite Used' - same
  - 'Aggregation Deleted' - same
  - 'Screen' - I have added connection_id to Screen events I encountered along the way + all non-modal Screens should have it. that means there might be modal Screen events that don't have connection_id. I can go through the rest, it's just extra work and I wasn't sure if it's necessary.
 
And these are general scope

  - 'Application Launched'
  - 'Application Restart Accepted'
  - 'Atlas Link Clicked'
  - 'Atlas Sign In Error'
  - 'Atlas Sign In Success'
  - 'Atlas Sign Out'
  - 'Autoupdate Accepted'
  - 'Autoupdate Disabled'
  - 'Autoupdate Dismissed'
  - 'Autoupdate Enabled'
  - 'My Queries Filter'
  - 'My Queries Sort'
  - 'My Queries Search'
  - 'Theme Changed'
  - 'Error Fetching Attributes'
  - 'Query History Favorite Copied' - this is from 'my_queries' screen
  - 'Aggregation Copied'
  - 'First Contentful Paint'
  - 'Largest Contentful Paint'
  - 'First Input Delay'
  - 'Cumulative Layout Shift'
  - 'Time to First Byte'
  - 'Guide Cue Dismissed'
  - 'Guide Cue Group Dismissed'
  - 'Connection Exported' - this actually seems to be a bulk operation
  - 'Connection Imported' - same as above
  - 'Keytar Secrets Migration Failed'
  - 'Signal Action Button Clicked'
  - 'Signal Closed'
  - 'Signal Link Clicked'
  - 'Signal Opened'
  - 'Signal Shown'

### Checklist
- [ ] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [ ] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [ ] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-telemetry/src/generic-track.ts [hunk 1]
```diff
@@ -55,10 +57,20 @@ export const createTrack = ({
         return;
       }
     }
-    sendTrack(event, properties);
+
+    if (typeof parameters === 'object' && connectionInfo) {
+      parameters.connection_id = connectionInfo.id;
+    }
+
+    // console.log('TRACK', {
```
Comment (mcasimir) [inline packages/compass-telemetry/src/generic-track.ts:65]: leftover?

Comment (paula-stacho) [reply packages/compass-telemetry/src/generic-track.ts]: yes, thank you!

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts [hunk 2]
```diff
@@ -94,6 +98,7 @@ export const changePipelineMode = (
     track('Editor Type Changed', {
       num_stages,
       editor_view_type: mapPipelineModeToEditorViewType(getState()),
+      connectionInfoAccess,
```
Comment (himanshusinghs) [inline packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts:101]:
```suggestion
      connectionInfoAccess.getCurrentConnectionInfo(),
```

File: packages/compass-indexes/src/components/create-index-modal/create-index-modal.tsx [hunk 3]
```diff
@@ -55,11 +57,16 @@ function CreateIndexModal({
 
   useTrackOnChange(
     (track: TrackFunction) => {
+      const connectionInfo = connectionInfoAccess.getCurrentConnectionInfo();
```
Comment (himanshusinghs) [inline packages/compass-indexes/src/components/create-index-modal/create-index-modal.tsx:60]: I noticed that in standard hooks you added connectionInfoAccess as a dependency but not here, maybe a miss?

File: packages/compass-indexes/src/components/search-indexes-modals/base-search-index-modal.tsx [hunk 4]
```diff
@@ -190,11 +192,16 @@ export const BaseSearchIndexModal: React.FunctionComponent<
   useTrackOnChange(
     (track: TrackFunction) => {
       if (isModalOpen) {
-        track('Screen', { name: `${mode}_search_index_modal` });
+        const connectionInfo = connectionInfoAccess.getCurrentConnectionInfo();
+        track('Screen', { name: `${mode}_search_index_modal` }, connectionInfo);
```
Comment (himanshusinghs) [inline packages/compass-indexes/src/components/search-indexes-modals/base-search-index-modal.tsx:196]: The same here, missing in the dependency?

File: packages/compass-query-bar/src/components/query-history/index.tsx [hunk 5]
```diff
@@ -40,13 +41,15 @@ const QueryHistory = ({
   onUpdateRecentChoosen,
 }: QueryHistoryProps) => {
   const [tab, setTab] = useState<QueryHistoryTab>('recent');
+  const connectionInfoAccess = useConnectionInfoAccess();
 
   useTrackOnChange(
     (track: TrackFunction) => {
+      const connectionInfo = connectionInfoAccess.getCurrentConnectionInfo();
```
Comment (himanshusinghs) [inline packages/compass-query-bar/src/components/query-history/index.tsx:48]: The same here, missing connectionInfoAccess in dependency?

File: packages/data-service/src/connect-mongo-client.spec.ts [hunk 6]
```diff
@@ -229,7 +229,7 @@ describe('connectMongoClient', function () {
         asyncHook.disable();
       });
 
-      it('should close ssh tunnel if the connection fails', async function () {
+      it.skip('should close ssh tunnel if the connection fails', async function () {
```
Comment (himanshusinghs) [inline packages/data-service/src/connect-mongo-client.spec.ts:232]:
```suggestion
      it('should close ssh tunnel if the connection fails', async function () {
```

Comment (himanshusinghs) [reply packages/data-service/src/connect-mongo-client.spec.ts:232]: Should we bring this back?

File: packages/databases-collections/src/components/collections.tsx [hunk 7]
```diff
@@ -54,11 +54,12 @@ const Collections: React.FunctionComponent<CollectionsListProps> = ({
...
     return isCompassInWritableMode && isInstanceWritable;
   }, [isCompassInWritableMode, isInstanceWritable]);
-  const { id: connectionId } = useConnectionInfo();
+  const connectionInfo = useConnectionInfo();
+  const { id: connectionId } = connectionInfo;
   const { openCollectionWorkspace } = useOpenWorkspace();
 
   useTrackOnChange((track: TrackFunction) => {
-    track('Screen', { name: 'collections' });
+    track('Screen', { name: 'collections' }, connectionInfo);
```
Comment (himanshusinghs) [inline packages/databases-collections/src/components/collections.tsx:62]: if we need to account for connectionInfo changing (or maybe we don't) then this probably should be part of dependency?

File: packages/databases-collections/src/components/databases.tsx [hunk 8]
```diff
@@ -112,7 +113,7 @@ const Databases: React.FunctionComponent<DatabasesProps> = ({
   }, [connectionId, _onCreateDatabaseClick]);
 
   useTrackOnChange((track: TrackFunction) => {
-    track('Screen', { name: 'databases' });
+    track('Screen', { name: 'databases' }, connectionInfo);
```
Comment (himanshusinghs) [inline packages/databases-collections/src/components/databases.tsx:116]: the same here?

**Review by himanshusinghs — APPROVED** [review]
Looking good. I left a few small suggestions and question.

File: packages/compass-e2e-tests/tests/collection-bulk-update.test.ts [hunk 9]
```diff
@@ -50,6 +50,10 @@ describe('Bulk Update', () => {
 
     // Check the telemetry
     const openedEvent = await telemetryEntry('Bulk Update Opened');
```
Comment (mcasimir) [inline packages/compass-e2e-tests/tests/collection-bulk-update.test.ts:52]:
```
const {connection_id, ...openedEvent} = await telemetryEntry('Bulk Update Opened');

expect(connection_id).to.exist;
expect(openedEvent).to.deep.equal({
      isUpdatePreviewSupported: true,
    });
```

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts (hunk 2 above)
Comment (paula-stacho) [reply packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts:101]: ah, I was wondering how did this pass TS, it was supposed to be as a third parameter, not inside 2nd one 🤦 thanks for spotting it!

File: packages/compass-indexes/src/components/create-index-modal/create-index-modal.tsx (hunk 3 above)
Comment (paula-stacho) [reply packages/compass-indexes/src/components/create-index-modal/create-index-modal.tsx:60]: thank you!

File: packages/data-service/src/connect-mongo-client.spec.ts (hunk 6 above)
Comment (paula-stacho) [reply packages/data-service/src/connect-mongo-client.spec.ts:232]: yes, thanks for the spot!

File: packages/compass-query-bar/src/components/query-history/index.tsx (hunk 5 above)
Comment (paula-stacho) [reply packages/compass-query-bar/src/components/query-history/index.tsx:48]: thank you! added a eslint rule to have a reminder next time.

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts (hunk 2 above)
Comment (himanshusinghs) [reply packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts]: ahh i see now

File: packages/compass-query-bar/src/components/query-history/index.tsx (hunk 5 above)
Comment (himanshusinghs) [reply packages/compass-query-bar/src/components/query-history/index.tsx:48]: amazing and thank you <3

File: packages/compass-e2e-tests/tests/collection-bulk-update.test.ts (hunk 9 above)
Comment (paula-stacho) [reply packages/compass-e2e-tests/tests/collection-bulk-update.test.ts:52]: so I checked the `browser.listenForTelemetryEvents` and it is doing a copy of the properties!

**Pull request merged**